teps/0079-tekton-catalog-support-tiers.md

```diff
@@ -100,14 +117,157 @@ requirements, and provenance attestations can be made to meet software supply ch
+  they must be resolved as soon as possible; the SLO is one week. The automated testing is discussed further [below](#automated-testing-and-dogfooding-1).
+1. The resource must be scanned for common vulnerabilities and exposures. If any issues are discovered, they must be 
+  patched or disclosed as soon as possible; the SLO is one week. The scanning for CVEs is discussed further [below](#image-scanning-for-common-vulnerabilities-and-exposures-cves-1).
+1. The resource must support verification as proposed in [TEP-0091: Verified Remote Resources][tep-0091]. 
+  For now, it must be published to a public OCI registry as a [Tekton Bundle][bundle] and signed by Tekton. 
+  In the future, it must be updated to support accepted proposal of TEP-0091 depending on the direction of 
+  [TEP-0060: Remote Resource Resolution][tep-0060]. The verification is discussed further [below](#verified-remote-resources-1).
+1. The resource must be well documented with all configuration options described and a working example provided.
+1. The resource must follow and demonstrate best practices, such as the [`Task` authoring recommendations][task-authoring-recommendations].
+
+#### Responsibilities
+
+There are two types of ownership: resource ownership and catalog ownership.
+
+The responsibilities of owning a resource include:
+
+1. Reviewing and approving all changes to the resource.
+1. Resolving any failures of tests validating the resource.
+1. Responding to and resolving any issues reported concerning the resource.
+1. Updating the resources to be compatible with new versions of Tekton.
+
+The responsibilities of owning a Catalog include:
+
+1. Reviewing and approving new resources contributed to the Catalogs.
+1. Maintaining the health of the testing infrastructure.
+1. Triaging issues and involving the owners of the affected resources.
+1. Ensuring the resources meet the quality standards of the Catalog.
+
+#### Design Evaluation
+
+The Catalogs will leverage existing structure and infrastructure that supports the existing `tektoncd/catalog`.
+Decoupling the support tiers into separate repositories makes it easy to distinguish the support tiers, 
+enforce the applicable quality standards, and leverage existing structure and infrastructure. 
+Moreover, reusing the existing Catalog at `tektoncd/catalog` minimizes effort to support the tiers.
+
+#### Alternatives
+
+##### 1. One Repository and Using Annotations
+
+Instead of creating new repository, add a new `tekton.dev/tier` annotation to the resource yaml file with the relevant 
+support tier, for example:
+
+```yaml
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.17.0"
+    tekton.dev/tags: image-build
+    tekton.dev/tier: official
+```
+
+The `tekton.dev/tier` annotation should be optional. If the annotation is not specified, `community` support tier is 
+meant by default. This will also provide backward compatibility to the existing resources in the Tekton Catalog.
+
+###### Design Evaluation
```

Comment:
Instead of editing the Task yaml, a metadata file that includes owners and support tier could work well here, as well as the Hub badges from the proposed solution. (I would expect most users search for Tasks in the Hub instead of in the Catalog repo.) I think it would still be possible here to apply different testing, security, etc requirements to tasks with different support levels indicated in metadata files.

I would prefer this solution for a few reasons:
- There's likely overlap between infrastructure required for the community catalog and official catalog, and having it in one repo would reduce infrastructure duplication.
- It would be easier to allow Community Tasks to opt into some of the additional testing/security requirements without having to become Official Tasks.
- Changing the label of a Task is easier than moving it between repos and doesn't result in a deprecated version.
- This strategy makes it easier to modify the support tiers in the future.

Reply:
Thank you for the review @lbernick! 🙏🏾 

> Instead of editing the Task yaml, a metadata file that includes owners and support tier could work well here

That's an interesting solution, but there's a challenge with this. The OWNERs file is at the resource level applying to all versions of the resource: `./{resource-type}/{resource-name}/OWNERS`. On the other hand, the initial versions of the resource would start at `Community` before transitioning to `Official`. This makes it hard to implement the suggested approach because we'd have to indicate the support tier and maintainers, of each version at the resource level. We could move this information to the version level but that would be backwards incompatible - we'd have to change the existing catalog organization contract. 

> There's likely overlap between infrastructure required for the community catalog and official catalog, and having it in one repo would reduce infrastructure duplication.

We plan to utilize the same infrastructure, including the testing infra, for both catalogs so there won't be much duplication. The hub is already set up to support multiple catalogs, [read more](https://github.com/tektoncd/community/blob/main/teps/0003-tekton-catalog-organization.md#the-hub-and-multiple-catalogs). Will address this concern in the TEP.

> It would be easier to allow Community Tasks to opt into some of the additional testing/security requirements without having to become Official Tasks.

To clarify, community resources can be tested, scanned or verified - it's just that it's not requirement for them. If they are tested, scanned or verified, we will surface that information in the Hub similarly to how we'd do it for official resources. Will add this clarification to the TEP as well. 

> Changing the label of a Task is easier than moving it between repos and doesn't result in a deprecated version.

We can provide tooling to make the migration from community to official easy. And we don't have to deprecate the community versions of the resource if users would find them useful, what I was getting at is indicating in some way that there's an official version of the resource elsewhere. 

> This strategy makes it easier to modify the support tiers in the future.

Agreed, but we can also provide more catalogs for new support tiers. We have also discussed supporting contributors/organizations defining their own catalogs and sharing them on the hub, as long as they meet the catalog contract and hub requirements, [read more](https://github.com/tektoncd/community/pull/170#discussion_r474920639).

Reply:
> Instead of editing the Task yaml, a metadata file that includes owners and support tier could work well here, as well as the Hub badges from the proposed solution. (I would expect most users search for Tasks in the Hub instead of in the Catalog repo.) I think it would still be possible here to apply different testing, security, etc requirements to tasks with different support levels indicated in metadata files.

+1 to what @jerop said. There is already OWNERS file located at `./{resource-type}/{resource-name}/OWNERS` for each resource and adding one extra metadata file would require a change to TEP-0003. Also please note that right now Tekton Hub's implementation is also done in such a way that in order to add a new support tier we need to add a new catalog repo in [config.yaml](https://github.com/tektoncd/hub/blob/main/config.yaml#L38-L43) and modify the `type` field over there.

 
> * There's likely overlap between infrastructure required for the community catalog and official catalog, and having it in one repo would reduce infrastructure duplication.

I don't think there will be any testing infrastructure duplication instead would help reduce the costs as only official catalog would be eligible for some special testing infra which community catalog won't be getting access to.


> * Changing the label of a Task is easier than moving it between repos and doesn't result in a deprecated version.

I agree but there can be a case where a task not eligible to be a part of official tier gets into a catalog and be a part of official tier.

Reply:
Would it work to change the OWNERS file to contain a mapping from version -> {support tier, maintainers}? This would be backwards incompatible, but the updates to the owners files could be easily automated, so we would not need to rely on Task owners to make the changes. We could also include a syntax to specify multiple versions at once to reduce duplication, e.g.
```
version: >= 0.1, <0.5
tier: community
maintainers: x@y.com

version: >=0.5
tier: official
maintainers: x@y.com
```
(inspiration here was dependency lockfiles.)

I didn't realize the infra was already set up to work with multiple repos and that community tasks can already have additional testing. Thanks for this info.

Reply:
Another downside to multiple repositories: if I'm looking for something, I now have to look for it in 2 places. As a user, I'm happier if I only have to look in one place.

Reply:
@vinamra28 thank you for sharing your perspectives! 

@lbernick clarified the points you raised in the TEP, and included the alternative you proposed :)

@bendory makes sense, but users usually go to the Tekton Hub - https://hub.tekton.dev/ - to look for resources instead of the Catalogs